### PR TITLE
Animation: Add guidelines for product animation

### PR DESCRIPTION
**Pull request by ayeshakmaz**
#### What changed?

Add animation principles, guidelines, and implementation details

#### Why?

Create initial guidelines around animation and motion

### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-4988)
- [TDD](https://coda.io/d/Product-Animation-Guidelines_dXH62hWG_Q-/Product-Animation-Guidelines_suius#_lupOS)

### Checklist

- [ ] Added unit and Flow Tests
- [ ] Added documentation + accessibility tests
- [ ] Verified accessibility: keyboard & screen reader interaction
- [ ] Checked dark mode, responsiveness, and right-to-left support
- [ ] Checked stakeholder feedback (e.g. Gestalt designers)

=

### CODE REVIEW

File: docs/pages/foundations/animation/guidelines.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+              <Markdown
+                text={`
+              Another common use case for animations is to show a transition between two states in the UI. For example, a search IconButton may expand into a SearchField when interacted with, or a Sheet may slide in from the side of the screen to draw attention to it. Transitional animations help draw the user's attention to particular pieces of the UI, ideally helping them complete a task.
+
+              Transitional animations should also follow an intuitive flow. For example, a drawer that appears from the bottom on mobile should slide up from the bottom of the screen and slide back down upon dismissal.
+              `}
+              />
+              <Box width={800}>
+                <SandpackExample
+                  code={transitionExample}
```
Comment (AlbertCarreras) [inline docs/pages/foundations/animation/guidelines.js:69]: this one doesnt load

File: docs/pages/foundations/animation/guidelines.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        <MainSection.Subsection
+          title="Easing and velocity"
+          description={`
+          The timing function of an animation should be determined by the property and object being animated.
+
+          Properties unrelated to motion (like color and opacity) will not likely benefit from easing and can be done linearly. For example, a background color of a button that expands or changes upon hover can happen linearly as it provides feedback to the user about their cursor position.
+
+          However, people experience motion non-linearly, so objects being moved can benefit from an easing timing function. For example, an object transitioning onto the screen from the right side may use ease-out as it slides into view to create visibility as fast as possible, and then ease-in upon its departure to create a smooth transition back to the main screen.
+
+          [Learn more about easing functions](https://www.smashingmagazine.com/2021/04/easing-functions-css-animations-transitions/#cubic-b%C3%A9zier-functions)
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/guidelines.js:163]: The placement of this link is a bit awkward. Can we include it in the preceding paragraph?

File: docs/pages/foundations/animation/guidelines.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+          `}
+        >
+          <SandpackExample
+            code={easeExample}
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/guidelines.js:167]: The "bounce" example is pretty quick — consider adding more time to the end of the animation so the bounce is more apparent

File: docs/pages/foundations/animation/guidelines.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        description={`
+        Animations should never cause a user harm or distract from a necessary task. However, inaccessible animations have the potential to cause seizures, motion sickness, or anxiety in users, so we must ensure we are creating accessible animations every time.
+
+        Below are 3 key considerations for accessibility. Note: these do not apply for “essential” animations, where [essential is defined](https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits-pause.html#essentialdef) as any animation where “if removed, would fundamentally change the information or functionality of the content, _and_ information and functionality cannot be achieved in another way that would conform.”
+`}
+      >
+        <MainSection.Subsection
+          title="Looping animations"
+          description={`
+          Ensure that any animation that is moving, blinking, scrolling, or auto-playing and lasts longer than 5 seconds has an affordance for pausing and resuming the animation or an option to hide the animation.
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/guidelines.js:186]: oxford comma 😭

File: docs/pages/foundations/animation/guidelines.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+          _Looping animations should be avoided whenever possible, as they can become distracting very quickly._
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/guidelines.js:188]: Should we not loop the above animation? 😅 (or provide a way to pause it?)

File: docs/pages/foundations/animation/guidelines.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+          Ensure that there are no flashing animations present, where flashing is defined as anything that flashes _more than three times in a one second period_.
+
+          Further reading: [WCAG 2.0 Three Flashes requirements](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-three-times.html)
+          `}
+          columns={2}
+        />
+        <MainSection.Subsection
+          title="Reduced motion user setting"
+          description={`
+          Any non-essential animations should automatically be disabled when a user enables the reduced motion setting on their device. We have provided a [useReducedMotion hook](https://gestalt.pinterest.systems/web/utilities/usereducedmotion) to make this easy to implement.
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/guidelines.js:206]: The hook is only for web, so we should indicate that.

Also: are there iOS/Android resources around reducing motion that we can link in addition to MDN?

File: docs/pages/foundations/animation/implementation.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    <Page title="Implementing animations">
+      <PageHeader
+        badge="pilot"
+        name="Implementing animations"
+        type="guidelines"
+        description="There are multiple ways to implement animations in product."
+      />
+      <MainSection
+        name="Gestalt"
+        description="Many of our components come with animations baked in! If you are implementing a custom animation, be sure to wrap the animation in the [useReducedMotion hook](https://gestalt.pinterest.systems/web/utilities/usereducedmotion) to ensure an accessible experience. This hook will automatically disable any contained animations when the user enables the Reduced Motion setting on their device."
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/implementation.js:21]: This is very web-specific. Perhaps this section should be "Gestalt on web"?

Also: `useReducedMotion` doesn't wrap anything, or do anything automatically. It returns a boolean that must be used by the developer in their logic to show/hide the animation.

File: docs/pages/foundations/animation/implementation.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      />
+      <MainSection
+        name="CSS and InlineStyle"
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/implementation.js:24]: this is also very web-specific

File: docs/pages/foundations/animation/implementation.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        description={`
+        Animations can be implemented using CSS via the InlineStyle package. See [this diff for an example](https://sourcegraph.pinadmin.com/phabricator.pinadmin.com/diffusion/P/pinboard/-/blob/webapp/app/common/react/components/WebMentorship/mentees/annisa/Profile.js?L24:5).
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/implementation.js:26]: I wouldn't recommend linking to a Web Mentorship file, as those aren't intended to be production-ready. There are plenty of other examples in the actual app we can use. Also, we should probably obfuscate the link behind a pinch link. (And we're not linking to a diff, but a file)

Also: InlineStyle isn't really a package (like an npm package), it's just a component. If we're linking to a usage in a file, we may want to link to [the component itself](https://phabricator.pinadmin.com/diffusion/P/browse/master/webapp/app/packages/InlineStyle.js).

File: docs/pages/foundations/animation/implementation.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+        Please don’t create new SCSS/CSS files (or add to existing ones). Please don’t attempt to target and modify internals of Gestalt components, as the feature could break without warning in the future. And when possible, please use any relevant [design tokens](https://gestalt.pinterest.systems/foundations/design_tokens) (e.g. color tokens instead of hex values) to ensure that your feature automatically gets any future changes to those token values.
+        `}
+      />
+      <MainSection
+        name="Lottie"
+        description="Lottie has been implemented on mobile and web platforms, and is best used for illustration animations. For more details about using Lottie, view our [internal documentation](https://coda.io/d/Lottie-Pinterest_daMxB3E_zkB/For-design_su4Je)."
```
Comment (dangerismycat) [inline docs/pages/foundations/animation/implementation.js:33]: This link should also be obfuscated

File: docs/docs-components/PrincipleItem.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
+// @flow strict
+import { Text, Box, Flex, Heading } from 'gestalt';
+import { type Node } from 'react';
+
+type PrincipleItemProps = {|
```
Comment (dangerismycat) [inline docs/docs-components/PrincipleItem.js:5]: nit: this is the only component here, so no need for namespacing, just `Props` will do

**Review by dangerismycat — COMMENTED** [review]
There's a lot of inconsistency around capitalization. We should reserve capitalizing ("Button") for only when we're referring specifically to the component, and use lowercase ("button") when discussing the type of element more generically. This especially matters for something like "button", where that could refer to both Button and IconButton, but consistency is important regardless.

Also: this presentation seems a bit off. I think we should either show the entire virtual phone, or just show the animation without the virtual phone. Cutting off the phone frame doesn't look that great:
Screen Shot 2022-10-26 at 4 15 32 PM

File: docs/pages/foundations/animation/principles.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        badge="pilot"
+        name="Animation principles"
+        type="guidelines"
+        description="Animations can be utilized to provide feedback, add delight, and educate users. When considering adding animations to a product, consider these guiding principles."
+      />
+      <MainSection name="Principles">
+        <Flex gap={12} alignContent="between" wrap>
+          <Flex.Item flex="grow" flexBasis="0%" minWidth={275} maxWidth="45%">
+            <PrincipleItem
+              color="yellow-caramellow-450"
```
[the remark this answers is not in the document]
Comment (ayeshakmaz) [reply docs/pages/foundations/animation/principles.js:22]: Nope, we decided to remove images from Principles

File: docs/pages/foundations/animation/guidelines.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        <MainSection.Subsection
+          title="Animating UI elements"
+          description="UI elements refers to controls on a surface, like a Card, Button, Sheet or Pin."
+        >
+          <Flex gap={8} direction="column" maxWidth={572}>
+            <Flex gap={2} direction="column">
+              <Heading size="300">Feedback</Heading>
+              <Markdown text="Animation is often used to provide feedback to the user that an action they intended to take was successful. For example, a Button may shrink when pressed or tapped to indicate to the user that the Button was successfully interacted with. Similarly, a loading spinner may appear after a user submits a form to help indicate to the user that the form was submitted successfully and a process is beginning. Another example could be a Card that grows slightly when hovered/focused to help indicate which Card a user is interacting with." />
+              <SandpackExample
+                code={feedbackExample}
```
[the remark this answers is not in the document]
Comment (ayeshakmaz) [reply docs/pages/foundations/animation/guidelines.js:51]: We can update this in a V2

File: docs/pages/foundations/animation/guidelines.js (hunk 5 above)
Comment (ayeshakmaz) [reply docs/pages/foundations/animation/guidelines.js:188]: hmm good point

File: docs/pages/foundations/animation/guidelines.js (hunk 6 above)
Comment (ayeshakmaz) [reply docs/pages/foundations/animation/guidelines.js:206]: Added for iOS, can't seem to find any docs on Android...

File: docs/pages/foundations/animation/principles.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (dangerismycat) [reply docs/pages/foundations/animation/principles.js:22]: Should we remove the `color` prop then?

File: docs/pages/foundations/animation/guidelines.js (hunk 6 above)
Comment (dangerismycat) [reply docs/pages/foundations/animation/guidelines.js]: I can't either, even on Material. That seems…odd 🤔

File: docs/pages/foundations/animation/principles.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (ayeshakmaz) [reply docs/pages/foundations/animation/principles.js:22]: Yeah, I was planning to refactor PrincipleItem in a separate PR. Color is required at the moment, so it'll be a bigger update in other places where this cmp is used

File: docs/pages/foundations/animation/guidelines.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dangerismycat) [reply docs/pages/foundations/animation/guidelines.js:84]: Brave! https://media.giphy.com/media/1SMnTcxxNUSA0/giphy.gif

**Review by dangerismycat — APPROVED** [review]

**Pull request merged**